### PR TITLE
WIP: Virtual funding protocol

### PR DESCRIPTION
**Pull request by andrewgordstewart**
This is a start towards the virtual-funding protocol, to get some things fleshed out.

### CODE REVIEW

File: packages/wallet/src/redux/protocols/virtual-funding/states.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  [JOINT_CHANNEL_DESCRIPTOR]: AdvanceChannelState;
+  [GUARANTOR_CHANNEL_DESCRIPTOR]: AdvanceChannelState;
+}
+export interface WaitForGuarantorFunding extends Base {
+  type: 'VirtualFunding.WaitForGuarantorFunding';
+  [INDIRECT_GUARANTOR_FUNDING_DESCRIPTOR]: 'UpdateConsensusState<G>';
+}
+export interface WaitForApplicationFunding extends Base {
+  type: 'VirtualFunding.WaitForApplicationFunding';
+  [INDIRECT_APPLICATION_FUNDING_DESCRIPTOR]: 'UpdateConsensusState<J>';
```
Comment (kerzhner) [inline packages/wallet/src/redux/protocols/virtual-funding/states.ts]: Also, what is the reason this is an array?

File: packages/wallet/src/redux/protocols/virtual-funding/readme.md [hunk 2]
```diff
@@ -68,22 +68,18 @@ sequenceDiagram
 
 ## State machine diagram
 
-In the following diagram, to move from `WaitForChannelPreparation`, the protocol needs to receive the `Prepared` success message from both `PrepareChannel(J)` and `PrepareChannel(G)`.
+In the following diagram, to move from `WaitForChannelPreparation`, the protocol needs to receive the `Prepared` success message from both `AdvanceChannel(J)` and `AdvanceChannel(G)`.
```
[the remark this answers is not in the document]
Comment (andrewgordstewart) [reply packages/wallet/src/redux/protocols/virtual-funding/readme.md]: Correct. You might want to optimize a bit and do 1 and 2 in parallel -- you just can't start 2 until you've learned the channelID from 1.

But for now, it's probably better to keep things simple and have it be completely linear.

File: packages/wallet/src/redux/protocols/virtual-funding/states.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+export const INDIRECT_APPLICATION_FUNDING_DESCRIPTOR = 'indirectApplicationFunding';
+export type SubstateDescriptor =
+  | typeof JOINT_CHANNEL_DESCRIPTOR
+  | typeof GUARANTOR_CHANNEL_DESCRIPTOR
+  | typeof INDIRECT_GUARANTOR_FUNDING_DESCRIPTOR
+  | typeof INDIRECT_APPLICATION_FUNDING_DESCRIPTOR;
+
+export interface WaitForChannelPreparation extends Base {
+  type: 'VirtualFunding.WaitForChannelPreparation';
+  [JOINT_CHANNEL_DESCRIPTOR]: AdvanceChannelState;
```
[the remark this answers is not in the document]
Comment (andrewgordstewart) [reply packages/wallet/src/redux/protocols/virtual-funding/states.ts]: It's not an array -- this interface has a property whose key is the value of `JOINT_CHANNEL_DESCRIPTOR`. This way, the same descriptor can be used as both the property key and in the `protocolLocator` string.

[the remark this answers is not in the document]
Comment (kerzhner) [reply packages/wallet/src/redux/protocols/virtual-funding/states.ts]: Ah gotcha. Looks like this is ECMAScript 2015 computed property syntax.

**Pull request merged**